### PR TITLE
Android CI improvements, improves #2892

### LINKED ISSUE #2892: Set up Android Github Action pipeline
Coming from https://github.com/mockito/mockito/issues/2890#issuecomment-1409200309:

> we started an effort to set up some basic Android tests for us: https://github.com/mockito/mockito/tree/main/subprojects/androidTest However, this stalled as we couldn't get it to work on CI. If anybody affected here would like to dedicate some time to prevent these kind of issues from happening, please take a look at the existing test suite and infrastructure and submit a PR to get it passing on CI.

The presence of such Github Action would have helped to prevent regressions like https://github.com/mockito/mockito/issues/2890. 

@TimvdLippe I will try to pull it off (I don't work with Android sadly), it seems like Github Actions have support for it.

### PR DESCRIPTION
**Pull request by TWiStErRob**
followup on #2893 and #2899, fixes #2892

## Changes
 * Lock in `emulator-build` for reproducibility and hopefully stability.
 * Add more `emulator-options` to make the emulator more stable.
 * Upload instrumentation test results as GHA artifact.
 * Upload logcat from device as GHA artifact.
 * Output relevant logcat to GHA logs in-line.
 * Simplify how the versions are determined.

## Investigation details

My first idea (finding) was cold-boot, but apparently that's default behavior via `--no-snapshot`.
 
Anyway, I had a look about recent failures, and found the below and made some change in hopes of a more stable CI:
   
 * <details><summary><code>Error: Timeout waiting for emulator to boot.</code></summary>
   
   [Job](https://github.com/mockito/mockito/actions/runs/4076245148/jobs/7023715242#step:4:981) 
   
   ```
     The process '/Users/runner/Library/Android/sdk/platform-tools/adb' failed with exit code 1
     /Users/runner/Library/Android/sdk/platform-tools/adb shell getprop sys.boot_completed
     adb: no devices/emulators found
     The process '/Users/runner/Library/Android/sdk/platform-tools/adb' failed with exit code 1
     /Users/runner/Library/Android/sdk/platform-tools/adb shell getprop sys.boot_completed
     adb: no devices/emulators found
     ...
   ```
   </details>
   
   Stumbled upon this issue https://github.com/ReactiveCircus/android-emulator-runner/issues/160 in one of the comments in the example CIs. I guess I agree that locking in a known-good version is better than be at the mercy of unstable "stable" releases. Google has blindsided us with a "no-repro" at https://issuetracker.google.com/issues/191799887.
   
   Fix: added explicit `emulator-build`

 * <details><summary><code>Unknown failure: cmd: Can't find service: package</code></summary>
   
   [Job](https://github.com/mockito/mockito/actions/runs/4106509220/jobs/7084834115#step:4:613)
   
   ```
   > Task :androidTest:connectedDebugAndroidTest
   Exception thrown during onBeforeAll invocation of plugin com.google.testing.platform.plugin.android.AndroidDevicePlugin.
   Failed to install APK(s): /Users/runner/work/mockito/mockito/subprojects/androidTest/build/outputs/apk/androidTest/debug/androidTest-debug-androidTest.apk
   Unknown failure: cmd: Can't find service: package
   com.android.ddmlib.InstallException: Unknown failure: cmd: Can't find service: package
   	at com.android.ddmlib.internal.DeviceImpl.installRemotePackage(DeviceImpl.java:1315)
   	at com.android.ddmlib.internal.DeviceImpl.installPackage(DeviceImpl.java:1141)
   	at com.android.tools.utp.plugins.deviceprovider.ddmlib.DdmlibAndroidDevice.installPackage(DdmlibAndroidDevice.kt)
   	at com.android.tools.utp.plugins.deviceprovider.ddmlib.DdmlibAndroidDeviceController$executeAsync$deferred$1.invokeSuspend(DdmlibAndroidDeviceController.kt:173)
   	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
   	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
   	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
   	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
   	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
   	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
   
   > Task :androidTest:connectedDebugAndroidTest FAILED
   ```
   </details>
   
   Strange error, it looks like the emulator is in a bad state. This could happen because the GitHub Actions runners might be shared between different jobs, and the emulator might be in a bad state when we try to use it. This sounds weird, because I'm sure GHA ensures a clean state, but not sure about it. Anyway, I haven't found any other suggestion than to kill the emulator and cold-boot it. Which is achieved by the default `-no-snapshot` flag. The only option I see here is to retry somehow.
   
   Fix: N/A so far, 🤞 `emulator-build` helps. 
   
 * <details><summary><code>Unknown failure: cmd: Failure calling service package: Broken pipe (32)</code></summary>
   
   [Job](https://github.com/mockito/mockito/actions/runs/4106655761/jobs/7085158611#step:4:496)
   
   ```
   > Task :androidTest:connectedDebugAndroidTest
   Exception thrown during onBeforeAll invocation of plugin com.google.testing.platform.plugin.android.AndroidDevicePlugin.
   Failed to install APK(s): /Users/runner/work/mockito/mockito/subprojects/androidTest/build/outputs/apk/debug/androidTest-debug.apk
   Unknown failure: cmd: Failure calling service package: Broken pipe (32)
   com.android.ddmlib.InstallException: Unknown failure: cmd: Failure calling service package: Broken pipe (32)
   	at com.android.ddmlib.internal.DeviceImpl.installRemotePackage(DeviceImpl.java:1315)
   	at com.android.ddmlib.internal.DeviceImpl.installPackage(DeviceImpl.java:1141)
   	at com.android.tools.utp.plugins.deviceprovider.ddmlib.DdmlibAndroidDevice.installPackage(DdmlibAndroidDevice.kt)
   	at com.android.tools.utp.plugins.deviceprovider.ddmlib.DdmlibAndroidDeviceController$executeAsync$deferred$1.invokeSuspend(DdmlibAndroidDeviceController.kt:173)
   	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
   	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
   	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
   	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
   	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
   	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
   
   
   > Task :androidTest:connectedDebugAndroidTest FAILED
   ```
   </details>
   
   Not a clue what's happening here, let's assume it's the "same" problem as above.
   Retry might help in this case, but it's quite complex, not sure that we want to adopt this.
   https://github.com/kiwix/kiwix-android/blob/develop/contrib/instrumentation.sh
   
   Fix: N/A so far, 🤞 `emulator-build` helps.
   
 * <details><summary><code>Mockito cannot mock this class: interface org.mockitousage.androidtest.BasicInterface.</code></summary>
   
   [Job](https://github.com/mockito/mockito/actions/runs/4106511182/jobs/7085124984#step:4:674)
   
   ```
   > Task :androidTest:connectedDebugAndroidTest
   Starting 6 tests on test(AVD) - 13
   
   org.mockitousage.androidtest.BasicInstrumentedTests > mockAndUseBasicClassWithVerify[test(AVD) - 13] FAILED 
   	org.mockito.exceptions.base.MockitoException:
   	Mockito cannot mock this class: interface org.mockitousage.androidtest.BasicInterface.
   
   org.mockitousage.androidtest.BasicInstrumentedTests > mockAndUseBasicClassUsingLocalMock[test(AVD) - 13] FAILED 
   	org.mockito.exceptions.base.MockitoException:
   	Mockito cannot mock this class: interface org.mockitousage.androidtest.BasicInterface.
   
   org.mockitousage.androidtest.BasicInstrumentedTests > mockAndUseBasicInterfaceUsingAnnotatedMock[test(AVD) - 13] FAILED 
   	org.mockito.exceptions.base.MockitoException:
   	Mockito cannot mock this class: interface org.mockitousage.androidtest.BasicInterface.
   
   org.mockitousage.androidtest.BasicInstrumentedTests > mockAndUseBasicClassUsingAnnotatedMock[test(AVD) - 13] FAILED 
   	org.mockito.exceptions.base.MockitoException:
   	Mockito cannot mock this class: interface org.mockitousage.androidtest.BasicInterface.
   
   org.mockitousage.androidtest.BasicInstrumentedTests > mockAndUseBasicInterfaceAndVerify[test(AVD) - 13] FAILED 
   	org.mockito.exceptions.base.MockitoException:
   	Mockito cannot mock this class: interface org.mockitousage.androidtest.BasicInterface.
   
   org.mockitousage.androidtest.BasicInstrumentedTests > mockAndUseBasicInterfaceUsingLocalMock[test(AVD) - 13] FAILED 
   	org.mockito.exceptions.base.MockitoException:
   	Mockito cannot mock this class: interface org.mockitousage.androidtest.BasicInterface.
   Tests on test(AVD) - 13 failed: There was 6 failure(s).
   
   > Task :androidTest:connectedDebugAndroidTest FAILED
   ```
   
   </details>
   
   Undiagnosable as per https://github.com/mockito/mockito/pull/2893#discussion_r1094978058. What we know: the emulator started, the Gradle build completed, and app deployed, and tests are discovered. At this point it's usually the production/test code that's wrong, but not reproducible locally.

   Fix: Added artifact for the test results.

---

Note: I went through all referenced CIs here:
https://github.com/ReactiveCircus/android-emulator-runner#who-is-using-android-emulator-runner
to find some ideas on how to improve the Mockito CI.
Most of them used verbatim runners as we've set up, but some (major ones) used some extra options.

Additional findings, that I ignored for now:
 * https://github.com/slackhq/keeper/blob/b2847d74cc9f80b275a56efbda94c310ff7b668b/.github/workflows/run_instrumentation_tests.sh#L7  
   Source: https://github.com/slackhq/keeper/commit/ae8638aa1005cf317333ee86d54882254896e896 in https://github.com/slackhq/keeper/pull/58
 * https://github.com/google/android-fhir/pull/1709; they deleted it, but have a backup infrastructure internally at Google.
 * https://github.com/dotanuki-labs/norris/pull/578; they have heavy UI tests/screenshot tests, and for performance moved to emulator.wtf

## Checklist

 - [x] Read the [contributing guide](https://github.com/mockito/mockito/blob/main/.github/CONTRIBUTING.md)
 - [x] PR should be motivated, i.e. what does it fix, why, and if relevant how
 - [x] If possible / relevant include an example in the description, that could help all readers
       including project members to get a better picture of the change
 - [x] Avoid other runtime dependencies
 - [x] Meaningful commit history ; intention is important please rebase your commit history so that each
       commit is meaningful and help the people that will explore a change in 2 years
 - [x] The pull request follows coding style
 - [x] Mention `Fixes #<issue number>` in the description _if relevant_
 - [ ] At least one commit should mention `Fixes #<issue number>` _if relevant_

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -121,8 +117,54 @@ jobs:
       with:
         arch: x86_64
         api-level: ${{ matrix.android-api }}
-        target: ${{ matrix.android-image-type }}
-        script: ./gradlew :androidTest:connectedCheck --no-daemon --no-build-cache
+        # Workaround for https://issuetracker.google.com/issues/267458959
```
Comment (reta) [inline .github/workflows/ci.yml:120]: It looks massive, is it worth the complexity? (vs building on `26` only)?

Comment (TWiStErRob) [reply .github/workflows/ci.yml:120]: btw, most of the complexity is to get the logcat displayed and saved

**Review by reta — APPROVED** [review]

**Review by TimvdLippe — APPROVED** [review]
This is indeed quite a bit of complexity for running a couple of Android tests. That said, if CI is more stable with this, then let's merge this as-is. Thanks for the investigation!

**Comment by wcandillon** [comment]
@TWiStErRob we are affected by the intermittent `Can't find service: package` error. We start the emulator build the android app (this can take up to 20min) and then try to install the app which fails 50% of the time. I am wondering if we should build first, start the emulator and then install to improve our chances of the emulator to not be reachable?

**Comment by TWiStErRob** [comment]
If a build is 20 minutes, I would build in a separate job, publish the APK as artifact and then download the artifact separately, or use Gradle cache to make sure the whole building running FROM-CACHE. Also according to docs the emulator startup time can be greatly reduced by caching, see [reactivecircus/android-emulator-runner](https://github.com/ReactiveCircus/android-emulator-runner#usage--examples)

**Pull request merged**